File: platform_src/ios/Elastos/Classes/AppManager.swift
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

import Foundation

class AppManager {
    private static var appManager: AppManager?;

    /** The internal message */
    static let MSG_TYPE_INTERNAL = 1;
    /** The internal return message. */
    static let MSG_TYPE_IN_RETURN = 2;
    /** The internal refresh message. */
    static let MSG_TYPE_IN_REFRESH = 3;

    /** The external message */
    static let MSG_TYPE_EXTERNAL = 11;
    /** The external launcher message */
    static let MSG_TYPE_EX_LAUNCHER = 12;
    /** The external install message */
    static let MSG_TYPE_EX_INSTALL = 13;
    /** The external return message. */
    static let MSG_TYPE_EX_RETURN = 14;

    let mainViewController: MainViewController;
    var viewControllers = [String: TrinityViewController]();

    let appsPath: String;
    let dataPath: String;
    let configPath: String;
    let tempPath: String;

    var curController: TrinityViewController?;

    let dbAdapter: ManagerDBAdapter;

    var appList: [AppInfo];
    var appInfos = [String: AppInfo]();
    var lastList = [String]();
    let installer: AppInstaller;

    private var currentLocale = "en";

    private var launcherInfo: AppInfo? = nil;

    var installUriList = [String]();
    var launcherReady = false;

    init(_ mainViewController: MainViewController) {
//        PermissionManager();
        
        self.mainViewController = mainViewController;
        appsPath = NSHomeDirectory() + "/Documents/apps/";
        dataPath = NSHomeDirectory() + "/Documents/data/";
        configPath = NSHomeDirectory() + "/Documents/config/";
        tempPath = NSHomeDirectory() + "/Documents/temp/";

        let fileManager = FileManager.default
        var first = false;
        if (!fileManager.fileExists(atPath: appsPath)) {
            do {
                try fileManager.createDirectory(atPath: appsPath, withIntermediateDirectories: true, attributes: nil)
                first = true;
            }
            catch let error {
                print("Make appsPath error: \(error)");
            }
        }

        if (!fileManager.fileExists(atPath: dataPath)) {
            do {
                try fileManager.createDirectory(atPath: dataPath, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make dataPath error: \(error)");
            }
        }

        if (!fileManager.fileExists(atPath: configPath)) {
            do {
                try fileManager.createDirectory(atPath: configPath, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make configPath error: \(error)");
            }
        }

        if (!fileManager.fileExists(atPath: tempPath)) {
            do {
                try fileManager.createDirectory(atPath: tempPath, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make tempPath error: \(error)");
            }
        }

        dbAdapter = ManagerDBAdapter(dataPath);
//        try! dbAdapter.clean();
        installer = AppInstaller(appsPath, dataPath, tempPath, dbAdapter);
        appList = try! dbAdapter.getAppInfos();

        if first {
            saveLauncherInfo();
            copyConfigFiles();
        }

        saveBuiltInAppInfos();
        refreashInfos();

        AppManager.appManager = self;
    }

    static func getShareInstance() -> AppManager {
        return AppManager.appManager!;
    }

    func saveLauncherInfo() {
        let launcherPath = getAbsolutePath("www/launcher");
        var path = launcherPath;

        let fileManager = FileManager.default;
        var ret = fileManager.fileExists(atPath: path + "/manifest.json");
        if (!ret) {
            path = path + "/assets";
            ret = fileManager.fileExists(atPath: path + "/manifest.json");
            guard ret else {
                print("Launcher error: no exist mainifest.json");
                return;
            }
        }

        do {
            let info = try installer.parseManifest(path + "/manifest.json", true)!;
            try installer.copyAssetsFolder(launcherPath, appsPath + info.app_id);
            info.built_in = true;
            try dbAdapter.addAppInfo(info);
        }
        catch let error {
            print("Copy launcher error: \(error)");
        }
    }

    func copyConfigFiles() {
        let path = getAbsolutePath("www/config");
        do {
            try installer.copyAssetsFolder(path, configPath);
        }
        catch let error {
            print("Copy configPath error: \(error)");
        }
    }

    func getLauncherInfo() -> AppInfo {
        if launcherInfo == nil {
            launcherInfo = try! dbAdapter.getLauncherInfo();
        }
        return launcherInfo!;
    }

    func refreashInfos() {
        appList = try! dbAdapter.getAppInfos();
        appInfos = [String: AppInfo]();
        for info in appList {
            appInfos[info.app_id] = info;
        }
    }

    func getAppInfo(_ id: String) -> AppInfo? {
        if (id == "launcher") {
            return getLauncherInfo();
        }
        else {
            return appInfos[id];
        }
    }

    func getAppInfos() -> [String: AppInfo] {
        return appInfos;
    }

    func getStartPath(_ info: AppInfo) -> String {
        if (!info.remote) {
            return getAppUrl(info) + info.start_url;
        }
        else {
            return info.start_url;
        }
    }

    func getAppPath(_ info: AppInfo) -> String {
        if (!info.remote) {
            return appsPath + info.app_id + "/";
        }
        else {
            let index = info.start_url.range(of: "/", options: .backwards)!.lowerBound;
            return String(info.start_url[info.start_url.startIndex ..< index])  + "/";
        }
    }

    func getAppUrl(_ info: AppInfo) -> String {
        var url = getAppPath(info);
        if (!info.remote) {
            url = "file://" + url;
        }
        return url;
    }

    func getDataPath(_ id: String) -> String {
        var appId = id;
        if (id == "launcher") {
            appId = getLauncherInfo().app_id;
        }
        return dataPath + appId + "/";
    }

    func getDataUrl(_ id: String) -> String {
        return "file://" + getDataPath(id);
    }

    func getTempPath(_ id: String) -> String {
        var appId = id;
        if (id == "launcher") {
            appId = getLauncherInfo().app_id;
        }
        return tempPath + appId + "/";
    }

    func getConfigPath() -> String {
        return configPath;
    }

    func getTempUrl(_ id: String) -> String {
        return "file://" + getTempPath(id);
    }
    
    func getIconPath(_ info: AppInfo) -> String {
        if (info.type == "url") {
            return appsPath + info.app_id + "/";
        }
        else {
            return getAppPath(info);
        }
    }

    func saveBuiltInAppInfos() {
        let path = getAbsolutePath("www/built-in") + "/";

        let fileManager = FileManager.default;
        let dirs = try? fileManager.contentsOfDirectory(atPath: path);
        guard dirs != nil else {
            return;
        }

        do {
            for dir in dirs! {
                var needInstall = true;
                for info in appList {
                    if info.app_id == dir {
                        needInstall = false;
                        break;
                    }
                }

                if (needInstall) {
                    try installer.copyAssetsFolder(path +  dir, appsPath + dir);
                    let info = try installer.getInfoByManifest(appsPath +  dir);
                    guard (info != nil || info!.app_id != "") else {
                        return;
                    }

                    info!.built_in = true;
                    try dbAdapter.addAppInfo(info!);
                }
            }
        } catch AppError.error(let err) {
            print(err);
        } catch let error {
            print(error.localizedDescription);
        }
    }

    func install(_ url: String) throws -> AppInfo? {
        let info = try! installer.install(self, url);
        if (info != nil) {
            refreashInfos();
            sendRefreshList("installed", info!);
        }
        return info;
    }

    func unInstall(_ id: String) throws {
        try close(id);
        try installer.unInstall(appInfos[id]);
        refreashInfos();
        sendRefreshList("unInstalled", appInfos[id]!);
    }

    func removeLastlistItem(_ id: String) {
        for (index, item) in lastList.enumerated() {
            if item == id {
                lastList.remove(at: index);
                break;
            }
        }
    }

    func switchContent(_ to: TrinityViewController, _ id: String) {
        mainViewController.switchController(from: curController!, to: to)

        removeLastlistItem(id);
        lastList.insert(id, at: 0);
    }

    func start(_ id: String) throws {
        var viewController = viewControllers[id]
        if viewController == nil {
            if (id == "launcher") {
                viewController = LauncherViewController();
            }
            else {
                let appInfo = appInfos[id];
                guard appInfo != nil else {
                    throw AppError.error("No such app!");
                }
                let appViewController = AppViewController(appInfo!, WhitelistFilter(appInfo!));
//                appViewController.setInfo(id, appInfo!);
                viewController = appViewController;
                sendRefreshList("started", appInfo!);

            }

            mainViewController.add(viewController!)
            viewControllers[id] = viewController;

            lastList.insert(id, at: 0);
        }
        else {
            if (curController != viewController) {
                switchContent(viewController!, id);
            }
        }

        curController = viewController
    }

    func close(_ id: String) throws {
        if (id == "launcher") {
            throw AppError.error("Launcher can't close!");
        }

        let info = appInfos[id];
        if (info == nil) {
            throw AppError.error("No such app!");
        }

        let viewController = viewControllers[id]
        if (viewController == nil) {
            return;
        }

        if (viewController == curController) {
            let id2 = lastList[1];
            let viewController2 = viewControllers[id2]
            if (viewController2 == nil) {
                throw AppError.error("RT inner error!");
            }
            switchContent(viewController2!, id2);
        }

        viewControllers[id] = nil;
        viewController!.remove();
        sendRefreshList("closed", info!);
    }


    func loadLauncher() throws {
        try start("launcher");
    }

    func setInstallUri(_ uri: String) {
        if launcherReady {
            sendInstallMsg(uri);
        }
        else {
            installUriList.append(uri);
        }
    }

    func isLauncherReady() -> Bool {
        return launcherReady;
    }

    func setLauncherReady() {
        launcherReady = true;

        for uri in installUriList {
            self.sendInstallMsg(uri);
        }
    }

    private func sendInstallMsg(_ uri: String) {
        let msg = "{\"uri\":\"" + uri + "\", \"dev\":\"false\"}";
        do {
            try sendMessage("launcher", AppManager.MSG_TYPE_EX_INSTALL, msg, "system");
        } catch let error {
            print("Send install message: " + msg + " error!");
        }
    }

    private func sendRefreshList(_ action: String, _ info: AppInfo ) {
        let msg = "{\"action\":\"" + action + "\", \"id\":\"" + info.app_id + "\", \"name\":\"" + info.name + "\"}";
        do {
            try sendMessage("launcher", AppManager.MSG_TYPE_IN_REFRESH, msg, "system");
        }
        catch let error {
            print("Send message: " + msg + " error!");
        }
   }

    func sendMessage(_ toId: String, _ type: Int, _ msg: String, _ fromId: String) throws {
        let viewController = viewControllers[toId]
        if (viewController != nil) {
            viewController!.basePlugin!.onReceive(msg, type, fromId);
        }
        else {
            throw AppError.error(toId + " isn't running!");
        }
    }

    func sendMessageToAll(_ type: Int, _ msg: String, _ fromId: String) {
        for id in viewControllers.keys {
            viewControllers[id]!.basePlugin!.onReceive(msg, type, fromId);
        }
    }

    func setCurrentLocale(_ code: String) {
        currentLocale = code;
        sendMessageToAll(AppManager.MSG_TYPE_IN_REFRESH,
                         "{\"action\":\"currentLocaleChanged\", \"code\":\""
                         + code + "\"}", "launcher");
    }

    func getCurrentLocale() -> String {
        return currentLocale;
    }

    func getPluginAuthority(_ id: String, _ plugin: String) -> Int {
        let info = appInfos[id];
        if (info != nil) {
            for pluginAuth in info!.plugins {
                if (pluginAuth.plugin == plugin) {
                    return pluginAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }

    func getUrlAuthority(_ id: String, _ url: String) -> Int {
        let info = appInfos[id];
        if (info != nil) {
            for urlAuth in info!.urls {
                if (urlAuth.url == url) {
                    return urlAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }

    func setPluginAuthority(_ id: String, _ plugin: String, _ authority: Int) throws {
        let info = appInfos[id];
        guard (info != nil) else {
            throw AppError.error("No such app!");
        }

        for pluginAuth in info!.plugins {
            if (pluginAuth.plugin == plugin) {
                try dbAdapter.updatePluginAuth(pluginAuth, authority);
                pluginAuth.authority = authority;
                sendRefreshList("authorityChanged", info!);
                return;
            }
        }
        throw AppError.error("The plugin isn't in list!");
    }

    func setUrlAuthority(_ id: String, _ url: String, _ authority: Int) throws {
        let info = appInfos[id];
        guard (info != nil) else {
            throw AppError.error("No such app!");
        }

        for urlAuth in info!.urls {
            if (urlAuth.url == url) {
                try dbAdapter.updateUrlAuth(urlAuth, authority);
                urlAuth.authority = authority;
                sendRefreshList("authorityChanged", info!);
                return;
            }
        }
        throw AppError.error("The url isn't in list!");
    }

    func runAlertPluginAuth(_ info: AppInfo, _ pluginName: String,
                            _ plugin: CDVPlugin,
                            _ command: CDVInvokedUrlCommand) {

        func doAllowHandler(alerAction:UIAlertAction) {
            try? setPluginAuthority(info.app_id, pluginName, AppInfo.AUTHORITY_ALLOW);
            plugin.execute(command);
            let result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
            result?.setKeepCallbackAs(false);
            plugin.commandDelegate?.send(result, callbackId:command.callbackId);
        }

        func doRefuseHandler(alerAction:UIAlertAction) {
            try? setPluginAuthority(info.app_id, pluginName, AppInfo.AUTHORITY_DENY);
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR,
                                         messageAs: "Plugin:'" + pluginName + "' have not run authority.");
            result?.setKeepCallbackAs(false);
            plugin.commandDelegate.send(result, callbackId: command.callbackId)
        }

        let alertController = UIAlertController(title: "Plugin authority request",
                message: "App:'" + info.name + "' request plugin:'" + pluginName + "' access authority.",
                preferredStyle: UIAlertController.Style.alert)
        let cancelAlertAction = UIAlertAction(title: "Refuse", style: UIAlertAction.Style.cancel, handler: doRefuseHandler)
        alertController.addAction(cancelAlertAction)
        let sureAlertAction = UIAlertAction(title: "Allow", style: UIAlertAction.Style.default, handler: doAllowHandler)
        alertController.addAction(sureAlertAction)

        self.mainViewController.present(alertController, animated: true, completion: nil)
    }

    func runAlertUrlAuth(_ info: AppInfo, _ url: String) {
        func doAllowHandler(alerAction:UIAlertAction) {
            try? setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_ALLOW);
        }

        func doRefuseHandler(alerAction:UIAlertAction) {
            try? setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_DENY);
        }

        let alertController = UIAlertController(title: "Url authority request",
                                                message: "App:'" + info.name + "' request url:'" + url + "' access authority.",
                                                preferredStyle: UIAlertController.Style.alert)
        let cancelAlertAction = UIAlertAction(title: "Refuse", style: UIAlertAction.Style.cancel, handler: doRefuseHandler)
        alertController.addAction(cancelAlertAction)
        let sureAlertAction = UIAlertAction(title: "Allow", style: UIAlertAction.Style.default, handler: doAllowHandler)
        alertController.addAction(sureAlertAction)
        self.mainViewController.present(alertController, animated: true, completion: nil)
    }

    func getAppIdList() -> [String] {
        var ret = [String]();
        for info in appList {
            ret.append(info.app_id);
        }
        return ret;
    }

    func getAppInfoList() -> [AppInfo] {
        return appList;
    }

    func getRunningList() -> [String] {
        var ret = [String]();
        for id in viewControllers.keys {
            ret.append(id);
        }
        return ret;
    }

    func getLastList() -> [String] {
        return lastList;
    }

}

File: platform_src/android/app/src/main/java/org/elastos/trinity/dapprt/AppManager.java
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

package org.elastos.trinity.dapprt;

import android.app.AlertDialog;
//import android.app.FragmentManager;
//import android.app.FragmentTransaction;
import android.content.DialogInterface;
import android.content.res.AssetManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;

import org.apache.cordova.CordovaWebView;
import org.elastos.trinity.runtime.R;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

public class AppManager {

    /** The internal message */
    public static final int MSG_TYPE_INTERNAL = 1;
    /** The internal return message. */
    public static final int MSG_TYPE_IN_RETURN = 2;
    /** The external launcher message */
    public static final int MSG_TYPE_EXTERNAL_LAUNCHER = 3;
    /** The external install message */
    public static final int MSG_TYPE_EXTERNAL_INSTALL = 4;
    /** The external return message. */
    public static final int MSG_TYPE_EX_RETURN = 5;

    public static AppManager appManager;
    private WebViewActivity activity;
    private WebViewFragment curFragment = null;
    ManagerDBAdapter dbAdapter = null;

    public String appsPath = null;
    public String dataPath = null;

    private AppInstaller installer;

    protected HashMap<String, CordovaWebView> appViews;
    protected LinkedHashMap<String, AppInfo> appInfos;
    private ArrayList<String> lastList = new ArrayList<String>();
    public AppInfo[] appList;

    private ArrayList<String>  installUriList = new ArrayList<String>();
    private boolean launcherReady = false;

    AppManager(WebViewActivity activity) {
        AppManager.appManager = this;
        this.activity = activity;

        appsPath = activity.getFilesDir() + "/apps/";
        dataPath = activity.getFilesDir() + "/data/";

        File destDir = new File(appsPath);
        if (!destDir.exists()) {
            destDir.mkdirs();
        }
        destDir = new File(dataPath);
        if (!destDir.exists()) {
            destDir.mkdirs();
        }

        dbAdapter = new ManagerDBAdapter(activity);
//        dbAdapter.clean();

        installer = new AppInstaller();
        installer.init(activity, dbAdapter, appsPath, dataPath);

        if (dbAdapter.helper.isCreatedTables) {
            saveBuiltInAppInfos();
        }

        refreashInfos();
    }


    private void refreashInfos() {
        appList = dbAdapter.getAppInfos();
        appInfos = new LinkedHashMap();
        for (int i = 0; i < appList.length; i++) {
            appInfos.put(appList[i].app_id, appList[i]);
        }
    }

    public AppInfo getAppInfo(String id) {
        return appInfos.get(id);
    }

    public HashMap<String, AppInfo> getAppInfos() {
        return appInfos;
    }

    public String getStartPath(AppInfo info) {
        return resetPath(getAppUrl(info), info.start_url);
    }

    public String getAppPath(AppInfo info) {
        return appsPath + info.app_id;
    }

    public String getAppUrl(AppInfo info) {
        if (info.built_in == 1) {
            return "file:///android_asset/www/built-in/" + info.app_id + "/";
        }
        else {
            return "file://" + appsPath + info.app_id + "/";
        }
    }

    public String getDataPath(AppInfo info) {
        return dataPath + info.app_id;
    }

    public String getDataUrl(AppInfo info) {
        return "file://" + dataPath + info.app_id + "/";
    }

    public String resetPath(String dir, String origin) {
        if (origin.indexOf("http://") != 0 && origin.indexOf("https://") != 0
                && origin.indexOf("file:///") != 0) {
            while (origin.startsWith("/")) {
                origin = origin.substring(1);
            }
            origin = dir + origin;
        }
        return origin;
    }

    public void saveBuiltInAppInfos(){
        AssetManager manager = activity.getAssets();
        try {
            String[] appdirs= manager.list("www/built-in");
            for (String appdir : appdirs) {
                InputStream inputStream = manager.open("www/built-in/" + appdir + "/manifest.json");
                AppInfo info = installer.parseManifest(inputStream);
                if (info == null) {
                    return;
                }

                info.built_in = 1;
                dbAdapter.addAppInfo(info);
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public AppInfo install(String url) {
        AppInfo info = installer.install(this, url);
        if (info != null) {
            refreashInfos();
        }
        return info;
    }

    public boolean unInstall(String id) {
        if (!close(id)) {
            return false;
        }

        boolean ret = installer.unInstall(appInfos.get(id));
        if (ret) {
            refreashInfos();
        }
        return ret;
    }

    private WebViewFragment findFragmentById(String id) {
        FragmentManager manager = activity.getSupportFragmentManager();
        List<Fragment> fragments = manager.getFragments();
        for (int i = 0; i < fragments.size(); i++) {
            WebViewFragment fragment = (WebViewFragment)fragments.get(i);
            if ((fragment != null) && (fragment.id.equals(id))) {
                return fragment;
            }
        }
        return null;
    }

    public void switchContent(WebViewFragment fragment, String id) {
        FragmentManager manager = activity.getSupportFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();
        if ((curFragment != null) && (curFragment != fragment)) {
            transaction.hide(curFragment);
        }
        if (curFragment != fragment) {
            if (!fragment.isAdded()) {
                transaction.add(R.id.content, fragment, id);
            } else if (curFragment != fragment) {
                transaction.setCustomAnimations(android.R.animator.fade_in, android.R.animator.fade_out)
                        .show(fragment);
            }
//            transaction.addToBackStack(null);
            transaction.commit();
            curFragment = fragment;
        }
        lastList.remove(id);
        lastList.add(0, id);
    }

    public boolean doBackPressed() {
        if (curFragment.id.equals("launcher")) {
            return true;
        }
        else {
            switchContent(findFragmentById("launcher"), "launcher");
            return false;
        }
    }

    public boolean start(String id) {
        if (!id.equals("launcher")) {
            AppInfo info = appInfos.get(id);
            if (info == null) {
                return false;
            }
        }

        WebViewFragment fragment = findFragmentById(id);
        if (fragment == null) {
            if (id.equals("launcher")) {
                fragment = LauncherViewFragment.newInstance();
            }
            else {
                fragment = AppViewFragment.newInstance(id);
            }
        }
        switchContent(fragment, id);

        return true;
    }

    public boolean close(String id) {
        if (id.equals("launcher")) {
            return false;
        }

        AppInfo info = appInfos.get(id);
        if (info == null) {
            return false;
        }

        FragmentManager manager = activity.getSupportFragmentManager();
        WebViewFragment fragment = findFragmentById(id);
        if (fragment == null) {
            return true;
        }

        if (fragment == curFragment) {
            String id2 = lastList.get(1);
            WebViewFragment fragment2 = findFragmentById(id2);
            if (fragment2 == null) {
                return false;
            }
            switchContent(fragment2, id2);
        }

//        if (fragment.appView != null) {
//            fragment.appView.handleDestroy();
//        }

        FragmentTransaction transaction = manager.beginTransaction();
        transaction.remove(fragment);
        transaction.commit();
        lastList.remove(id);

        return true;
    }

    public boolean loadLauncher() {
        return start("launcher");
    }

    public void setInstallUri(String uri) {
        if (uri == null) return;

        if (launcherReady) {
            sendMessage("launcher", MSG_TYPE_EXTERNAL_INSTALL, uri, "system");
        }
        else {
            installUriList.add(uri);
        }
    }

    public boolean isLauncherReady() {
        return launcherReady;
    }

    public void setLauncherReady() {
        launcherReady = true;

        for (int i = 0; i < installUriList.size(); i++) {
            String uri = installUriList.get(i);
            sendMessage("launcher", MSG_TYPE_EXTERNAL_INSTALL, uri, "system");
        }
    }


    public boolean sendMessage(String toId, int type, String msg, String fromId) {
        FragmentManager manager = activity.getSupportFragmentManager();
        WebViewFragment fragment = (WebViewFragment)manager.findFragmentByTag(toId);
        if (fragment != null) {
            fragment.basePlugin.onReceive(msg, type, fromId);
            return true;
        }
        else {
            return false;
        }
    }

    public int getPluginAuthority(String id, String plugin) {
        AppInfo info = appInfos.get(id);
        if (info != null) {
            for (AppInfo.PluginAuth pluginAuth : info.plugins) {
                if (pluginAuth.plugin.equals(plugin)) {
                    return pluginAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }

    public int getUrlAuthority(String id, String url) {
        AppInfo info = appInfos.get(id);
        if (info != null) {
            for (AppInfo.UrlAuth urlAuth : info.urls) {
                if (urlAuth.url.equals(url)) {
                    return urlAuth.authority;
                }
            }
        }
        return AppInfo.AUTHORITY_NOEXIST;
    }

    public boolean setPluginAuthority(String id, String plugin, int authority) {
        AppInfo info = appInfos.get(id);
        if (info != null) {
            dbAdapter.updatePluginAuth(info.tid, plugin, authority);
            for (AppInfo.PluginAuth pluginAuth : info.plugins) {
                if (pluginAuth.plugin.equals(plugin)) {
                    pluginAuth.authority = authority;
                    return true;
                }
            }
        }
        return false;
    }

    public boolean setUrlAuthority(String id, String url, int authority) {
        AppInfo info = appInfos.get(id);
        if (info != null) {
            int count = dbAdapter.updateURLAuth(info.tid, url, authority);
            if (count > 0) {
                for (AppInfo.UrlAuth urlAuth : info.urls) {
                    if (urlAuth.url.equals(url)) {
                        urlAuth.authority = authority;
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private static void print(String msg) {
        String name = Thread.currentThread().getName();
        System.out.println(name + ": " + msg);
    }


    private class LockObj {
        int authority = AppInfo.AUTHORITY_NOINIT;
    }

    public synchronized int runAlertPluginAuth(AppInfo info, String plugin, int originAuthority) {
        LockObj lock = new LockObj();
        lock.authority = originAuthority;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                alertPluginAuth(info, plugin, lock);
            }
        });
        try {
            synchronized (lock) {
                if (lock.authority == originAuthority) {
                    lock.wait();
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
            return originAuthority;
        }
        return lock.authority;
    }

    public void alertPluginAuth(AppInfo info, String plugin, LockObj lock) {
        AlertDialog.Builder ab = new AlertDialog.Builder(activity);
        ab.setTitle("Plugin authority request");
        ab.setMessage("App:'" + info.name + "' request plugin:'" + plugin + "' access authority.");
        ab.setIcon(android.R.drawable.ic_dialog_info);


        ab.setPositiveButton("Allow", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
//                setPluginAuthority(info.app_id, plugin, AppInfo.AUTHORITY_ALLOW);
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_ALLOW;
                    lock.notify();
                }
            }
        });
        ab.setNegativeButton("Refuse", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                setPluginAuthority(info.app_id, plugin, AppInfo.AUTHORITY_DENY);
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_DENY;
                    lock.notify();
                }
            }
        });
        ab.show();
    }

    public synchronized int runAlertUrlAuth(AppInfo info, String url, int originAuthority) {
        LockObj lock = new LockObj();
        lock.authority = originAuthority;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                alertUrlAuth(info, url, lock);
            }
        });
        try {
            synchronized (lock) {
                if (lock.authority == originAuthority) {
                    lock.wait();
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
            return originAuthority;
        }
        return lock.authority;
    }

    public void alertUrlAuth(AppInfo info, String url, LockObj lock) {
        AlertDialog.Builder ab = new AlertDialog.Builder(activity);
        ab.setTitle("Url authority request");
        ab.setMessage("App:'" + info.name + "' request url:'" + url + "' access authority.");
        ab.setIcon(android.R.drawable.ic_dialog_info);
        ab.setPositiveButton("Allow", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_ALLOW);
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_ALLOW;
                    lock.notify();
                }
            }
        });
        ab.setNegativeButton("Refuse", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                setUrlAuthority(info.app_id, url, AppInfo.AUTHORITY_DENY);
                synchronized (lock) {
                    lock.authority = AppInfo.AUTHORITY_DENY;
                    lock.notify();
                }
            }
        });
        ab.show();
    }

    public String[] getAppList() {
        String[] ids = new String[appList.length];
        for (int i = 0; i < appList.length; i++) {
            ids[i] = appList[i].app_id;
        }
        return ids;
    }

    public String[] getRunningList() {
        FragmentManager manager = activity.getSupportFragmentManager();
        List<Fragment> fragments = manager.getFragments();
        String[] ids = new String[fragments.size()];

        for (int i = 0; i < fragments.size(); i++) {
            WebViewFragment fragment = (WebViewFragment)fragments.get(i);
            ids[i] = fragment.id;
        }
        return ids;
    }

    public String[] getLastList() {
        String[] ids = new String[lastList.size()];
        return lastList.toArray(ids);
    }
}
